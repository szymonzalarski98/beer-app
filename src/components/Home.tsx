import React from 'react';
import { SafeAreaView, Text, FlatList, StyleSheet, ActivityIndicator } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { FILTERS_VISIBILITY } from '../actions/types';
import { deviceWidth } from '../styles/helpers.style';
import theme from '../styles/theme.style';
import Filters from './filters/Filters';
import Product from './Product';
import Separator from './Separator';
interface Product {
  name: String;
  image_url: String;
  id: number;
  abv: number;
  ibu: number;
};

interface Props {
  data: Product [];
  isLoading: boolean;
  getData: Function;
  isEnd: boolean;
  handleFilters: Function;
};

const Home: React.SFC<Props> = (props) => {
  const dispatch = useDispatch();
  const filtersVisibility = useSelector((state: any) => state.filtersVisibility);
  const { data, isLoading, getData, isEnd, handleFilters } = props;
  let onEndReachedCalledDuringMomentum: boolean = true;
  return (
    <SafeAreaView style={styles.container}>
      <Text style={[theme.title, styles.title]}>The list of beers</Text>
      {
        filtersVisibility
        ?
        <Filters handleFilters={handleFilters} />
        :
        <Text
          style={[theme.text, styles.filters]}
          onPress={() => dispatch({ type: FILTERS_VISIBILITY, payload: true })}
        >
          Filters
        </Text>
      }
      <FlatList
        data={data}
        renderItem={({ item }) => <Product id={item.id} name={item.name} image={item.image_url} abv={item.abv} ibu={item.ibu} />}
        ListEmptyComponent={() => <Text style={[theme.text, { marginLeft: deviceWidth * 0.05 }]}>No results</Text>}
        ItemSeparatorComponent={() => <Separator />}
        keyExtractor={item => item.id.toString()}
        onEndReachedThreshold={0.5}
        onEndReached={({ distanceFromEnd }) => {
          if (!onEndReachedCalledDuringMomentum) {
            getData();
            onEndReachedCalledDuringMomentum = true;
          }
        }}
        onMomentumScrollBegin={() => onEndReachedCalledDuringMomentum = false}
        ListFooterComponent={() => !isEnd ? (isLoading && <ActivityIndicator style={styles.activityIndicator} />) : <Text style={[theme.text, { textAlign: 'center' }]}>End.</Text>}
        testID="products-list"
      />
    </SafeAreaView>
  );
};

export default Home;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  title: {
    textAlign: 'center',
    marginVertical: 10
  },
  activityIndicator: {
    marginBottom: 5,
  },
  filters: {
    textAlign: 'right',
    marginRight: deviceWidth * 0.05,
    fontWeight: 'bold',
  }
});