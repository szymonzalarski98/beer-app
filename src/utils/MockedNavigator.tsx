import React, { ReactNode } from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';
interface Props {
  component: ReactNode,
  params?: {}
}

const Stack = createStackNavigator();
const MockedNavigator: React.SFC<Props> = ({component, params = {}}) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          name="MockedScreen"
          component={component}
          initialParams={params}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default MockedNavigator;