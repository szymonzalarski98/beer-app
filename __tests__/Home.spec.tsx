import React from 'react';
import { render, waitForElement, act } from 'react-native-testing-library';
import axios from 'axios';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import MockedNavigator from '../src/utils/MockedNavigator';
import Home from '../src/components/Home';

const data = [
  {
    id: 1,
    name: 'Buzz',
    abv: 4.4,
    ibu: 4.5,
    image_url: 'https://images.punkapi.com/v2/1.png',
  },
  {
    id: 2,
    name: 'Trashy Blonde',
    abv: 4.4,
    ibu: 4.1,
    image_url: 'https://images.punkapi.com/v2/2.png',
  },
];

jest.mock('axios');

const mockStore = configureMockStore();

describe('Test Home component', () => {
  it ('Should render Product component two times', async () => {
    axios.get.mockResolvedValue({
      data
    });
    const HomeComponent = () =>  <Home data={data} isLoading={false} isEnd={false} getData={() => {}} handleFilters={() => {}} />;
    const store = mockStore({
      data: [],
      isLoading: false,
      getData: () => {},
      isEnd: true,
      handleFilters: () => {}
    });
    const { baseElement, getByText, getByTestId } = render(
      <Provider store={store}>
        <MockedNavigator component={HomeComponent} />
      </Provider>
    );
    await act( async () => await waitForElement(() => getByTestId('products-list')));
    expect(getByText('The list of beers')).toBeTruthy();
    expect(baseElement).toMatchSnapshot();
  });
});