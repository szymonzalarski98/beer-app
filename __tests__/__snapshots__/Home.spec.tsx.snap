// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Home component Should render Product component two times 1`] = `undefined`;
